tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,325 @@
+/// \file RClusterPool.cxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2020-03-11
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RClusterPool.hxx>
+#include <ROOT/RNTupleDescriptor.hxx>
+#include <ROOT/RPageStorage.hxx>
+
+#include <TError.h>
+
+#include <algorithm>
+#include <chrono>
+#include <future>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <utility>
+
+bool ROOT::Experimental::Detail::RClusterPool::RInFlightCluster::operator <(const RInFlightCluster &other) const
+{
+   if (fClusterId < other.fClusterId)
+      return true;
+   if (fClusterId > other.fClusterId)
+      return false;
+   if (fColumns.size() < other.fColumns.size())
+      return true;
+   if (fColumns.size() > other.fColumns.size())
+      return false;
+   for (auto itr1 = fColumns.begin(), itr2 = other.fColumns.begin(); itr1 != fColumns.end(); ++itr1, ++itr2) {
+      if (*itr1 == *itr2)
+         continue;
+      return *itr1 < *itr2;
+   }
+   // *this == other
+   return false;
+}
+
+ROOT::Experimental::Detail::RClusterPool::RClusterPool(RPageSource *pageSource, unsigned int size)
+   : fPageSource(pageSource)
+   , fPool(size, nullptr)
+   , fThreadIo(&RClusterPool::ExecLoadClusters, this)
+{
+   R__ASSERT(size > 0);
+   fWindowPre = 0;
+   fWindowPost = size;
+   // Large pools maintain a small look-back window together with the large look-ahead window
+   while ((1u << fWindowPre) < (fWindowPost - (fWindowPre + 1))) {
+      fWindowPre++;
+      fWindowPost--;
+   }
+}
+
+ROOT::Experimental::Detail::RClusterPool::~RClusterPool()
+{
+   {
+      // Controlled shutdown of the I/O thread
+      std::unique_lock<std::mutex> lock(fLockWorkQueue);
+      fWorkQueue.emplace(RWorkItem());
+      fCvHasWork.notify_one();
+   }
+   fThreadIo.join();
+}
+
+void ROOT::Experimental::Detail::RClusterPool::ExecLoadClusters()
+{
```

Comment:
Is there exactly one or more than one of those in flight at a time?

Reply:
Yes, there is exactly one I/O thread for every RClusterPool object.  I added that information to the method documentation